bazaarvoice/emodb
# Permission subset checks when creating roles

## Github Issue #

[63](https://github.com/bazaarvoice/emodb/issues/63)

## What Are We Doing Here?

This pull request contains the following overarching changes, also each of these contains several smaller details to make the overall implementation possible:

1. Added ability to determine if one `Condition` is a subset of another `Condition`.  For example, `like("ge*us")` is a subset of `like("g*s")`.
2. Expanded permission implication to include open-ended conditions.  For example, `sor|if(in("read","update","create_table","drop_table"))|if({..,"client":+})` implies `sor|if(in("read","update"))|if({..,"client":"testclient"})`, since for all permissions _p_ if the latter implies _p_ the former also implies _p_.
3. Enforced that when addition permissions to a role the calling use must have permission to perform the permissions he is granting.

## How to Test and Verify

1. Check out this PR
2. Start EmoDB locally
3. Run the following sequence and verify similar results:

```
$ curl -XPOST 'localhost:8081/tasks/role?action=update&role=delegation&permit=system|manage_roles&permit=sor|read|*&APIKey=local_admin'
Role updated.
delegation has 2 permissions
- sor|read|*
- system|manage_roles
$ curl -XPOST 'localhost:8081/tasks/api-key?action=create&owner=test&description=test&role=delegation&APIKey=local_admin'
API key: md4bastpqptnob9buiewz3g49kfvzi52ispfsjg32a2rfmyx

Warning:  This is your only chance to see this key.  Save it somewhere now.
$ curl -XPOST 'localhost:8081/tasks/role?action=update&role=illegal&permit=sor|*&APIKey=md4bastpqptnob9buiewz3g49kfvzi52ispfsjg32a2rfmyx'
You do not has sufficient permissions to grant the following permission(s):
- sor|*
Please remove or rewrite the above permission(s) constrained within your permissions
```

## Risk

The main risk associated with this ticket is potentially breaking the ability to create and update roles or a regression on evaluating permissions.

### Level 

Somewhere between `Medium` and `High`.  This change affects core Emo functionality -- security -- but does not put data at risk.

### Required Testing

`Manual`

### Risk Summary

In addition to the risks mentioned above and in the corresponding ticket, the biggest risk is in an incorrect implementation of condition subset and permission implication checks.  Any issues where a user can grant wider permissions than he has is a potential security hole.

## Code Review Checklist


build and run.

we inevitably come back to this code it will only take hours to figure out, not
days.

a victim of rampaging consistency, and should be using this course of action. 
We don't have coding standards out yet for this project, so please make sure to address any feedback regarding STYLE so the codebase remains consistent.
